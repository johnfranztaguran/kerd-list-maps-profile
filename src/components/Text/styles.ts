import { StyleSheet, TextStyle } from 'react-native';
import { themes, dimens, colors, fonts } from 'constant';

type Styles = {
  text: TextStyle;
  regular: TextStyle;
  bold: TextStyle;
  semibold: TextStyle;
  medium: TextStyle;
  light: TextStyle;
  center: TextStyle;
  right: TextStyle;
  primary: TextStyle;
  secondary: TextStyle;
  gray1: TextStyle;
  gray2: TextStyle;
  black: TextStyle;
  white: TextStyle;
  h1: TextStyle;
  h2: TextStyle;
  h3: TextStyle;
  h4: TextStyle;
  h5: TextStyle;
  body: TextStyle;
  subtitle: TextStyle;
  caption: TextStyle;
  label: TextStyle;
};

const styles: Styles = StyleSheet.create<Styles>({
  // default style
  text: {
    fontSize: dimens.fontSize.size_12,
    color: colors.black,
  },
  // variations
  regular: {
    fontWeight: 'normal',
  },
  bold: {
    fontWeight: 'bold',
  },
  semibold: {
    fontWeight: '500',
  },
  medium: {
    fontWeight: '500',
  },
  light: {
    fontWeight: '200',
  },
  // position
  center: { textAlign: 'center' },
  right: { textAlign: 'right' },
  // colors
  primary: { color: themes.colors.primary },
  secondary: { color: themes.colors.secondary },
  gray1: { color: colors.gray1 },
  gray2: { color: colors.gray2 },
  black: { color: colors.black },
  white: { color: colors.white },
  // fonts
  h1: {
    fontSize: themes.fonts.h1,
    fontFamily: fonts.OpenSansRegular,
    lineHeight: 28,
  },
  h2: {
    fontSize: themes.fonts.h2,
    fontFamily: fonts.OpenSansRegular,
    lineHeight: 26,
  },
  h3: {
    fontSize: themes.fonts.h3,
    fontFamily: fonts.OpenSansRegular,
    lineHeight: 23,
  },
  h4: {
    fontSize: themes.fonts.h4,
    fontFamily: fonts.OpenSansRegular,
    lineHeight: 21,
  },
  h5: {
    fontSize: themes.fonts.h5,
    fontFamily: fonts.OpenSansRegular,
    lineHeight: 19,
  },
  body: {
    fontSize: themes.fonts.body,
    fontFamily: fonts.OpenSansRegular,
    lineHeight: 16,
  },
  subtitle: {
    fontSize: themes.fonts.subtitle,
    fontFamily: fonts.OpenSansRegular,
    lineHeight: 14,
  },
  caption: {
    fontSize: themes.fonts.caption,
    fontFamily: fonts.OpenSansRegular,
    lineHeight: 12,
  },
  label: {
    fontSize: themes.fonts.label,
    fontFamily: fonts.OpenSansRegular,
    lineHeight: 9,
  },
});

export default styles;
