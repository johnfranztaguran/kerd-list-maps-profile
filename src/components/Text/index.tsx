import React from 'react';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { TextStyle, TextProps } from 'react-native';
import { Text } from 'react-native-paper';
import { isNil } from 'ramda';
import { testIdProps } from 'utils/testAutomation';
import { strings } from 'constant';
import styles from './styles';

type Props = {
  text: string;
  h1?: boolean;
  h2?: boolean;
  h3?: boolean;
  h4?: boolean;
  h5?: boolean;
  body?: boolean;
  subtitle?: boolean;
  caption?: boolean;
  label?: boolean;
  size?: number;
  transform?: 'capitalize' | 'uppercase' | 'lowercase' | 'none';
  align?: 'auto' | 'center' | 'right' | 'left' | 'justify';
  regular?: boolean;
  bold?: boolean;
  semibold?: boolean;
  medium?: boolean;
  light?: boolean;
  weight?: 'bold' | 'normal' | '100' | '200' | '300' | '400' | '500' | '600' | '700' | '800' | '900' // prettier-ignore
  center?: boolean;
  right?: boolean;
  spacing?: number;
  height?: number;
  color?: string;
  primary?: boolean;
  secondary?: boolean;
  gray1?: boolean;
  gray2?: boolean;
  black?: boolean;
  white?: boolean;
  style?: TextStyle | TextStyle[];
  decoration?: 'none' | 'line-through' | 'underline' | 'underline line-through';
};

const Typography = (props: Props & TextProps): JSX.Element => {
  const {
    text,
    h1,
    h2,
    h3,
    h4,
    h5,
    body,
    subtitle,
    caption,
    label,
    size,
    transform,
    align,
    // styling
    regular,
    bold,
    semibold,
    medium,
    light,
    weight,
    center,
    right,
    spacing, // letter-spacing
    height, // line-height
    // colors
    color,
    primary,
    secondary,
    gray1,
    gray2,
    black,
    white,
    style,
    decoration,
    ...otherProps
  } = props;

  const textStyles = [
    styles.text,
    h1 && styles.h1,
    h2 && styles.h2,
    h3 && styles.h3,
    h4 && styles.h4,
    h5 && styles.h5,
    body && styles.body,
    subtitle && styles.subtitle,
    caption && styles.caption,
    label && styles.label,
    !isNil(size) && { fontSize: size },
    transform && { textTransform: transform },
    align && { textAlign: align },
    !isNil(height) && { lineHeight: height },
    !isNil(spacing) && { letterSpacing: spacing },
    weight && { fontWeight: weight },
    regular && styles.regular,
    bold && styles.bold,
    semibold && styles.semibold,
    medium && styles.medium,
    light && styles.light,
    center && styles.center,
    right && styles.right,
    !isNil(color) && { color },
    // color shortcuts
    primary && styles.primary,
    secondary && styles.secondary,
    gray1 && styles.gray1,
    gray2 && styles.gray2,
    black && styles.black,
    white && styles.white,
    !isNil(decoration) && { textDecorationLine: decoration },
    style, // rewrite predefined styles
  ];

  return (
    <Text
      style={textStyles}
      {...(otherProps as TextProps)}
      {...testIdProps(strings.testIdText.prefixTxt + text)}>
      {text}
    </Text>
  );
};

export default Typography;
