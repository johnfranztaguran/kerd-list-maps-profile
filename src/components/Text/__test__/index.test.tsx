import React from 'react';
import { render } from '@testing-library/react-native';
import Text from '../index';

const createTextProps = (props: object) => ({
  ...props,
});

describe('Component: Text', () => {
  it('Should render a <Headline1 />', () => {
    const props: any = createTextProps({ text: 'Test h1', h1: true });
    const { baseElement } = render(<Text {...props} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('Should render a <Body />', () => {
    const props: any = createTextProps({ text: 'Test body', body: true });
    const { baseElement } = render(<Text {...props} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('Should render a <Subtitle />', () => {
    const props: any = createTextProps({
      text: 'Test subtitle',
      subtitle: true,
    });
    const { baseElement } = render(<Text {...props} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('Should render a <Caption />', () => {
    const props: any = createTextProps({
      text: 'Test caption',
      caption: true,
    });
    const { baseElement } = render(<Text {...props} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('Should render a <Label />', () => {
    const props: any = createTextProps({
      text: 'Test label',
      label: true,
    });
    const { baseElement } = render(<Text {...props} />);
    expect(baseElement).toMatchSnapshot();
  });
});
