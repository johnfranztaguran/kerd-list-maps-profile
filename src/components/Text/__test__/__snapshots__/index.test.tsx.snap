// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Text Should render a <Body /> 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Text
      accessibilityLabel="txt-test-body"
      accessible={true}
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "System",
            "fontWeight": "400",
            "textAlign": "left",
          },
          Array [
            Object {
              "color": "#000000",
              "fontSize": 12,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            Object {
              "fontFamily": "OpenSans",
              "fontSize": 14,
              "lineHeight": 16,
            },
            undefined,
            undefined,
            undefined,
            false,
            undefined,
            undefined,
            false,
            false,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            false,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
          ],
        ]
      }
      testID="txt-test-body"
    >
      Test body
    </Text>
  </View>
</View>
`;

exports[`Component: Text Should render a <Caption /> 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Text
      accessibilityLabel="txt-test-caption"
      accessible={true}
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "System",
            "fontWeight": "400",
            "textAlign": "left",
          },
          Array [
            Object {
              "color": "#000000",
              "fontSize": 12,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            Object {
              "fontFamily": "OpenSans",
              "fontSize": 10,
              "lineHeight": 12,
            },
            undefined,
            false,
            undefined,
            undefined,
            false,
            false,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            false,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
          ],
        ]
      }
      testID="txt-test-caption"
    >
      Test caption
    </Text>
  </View>
</View>
`;

exports[`Component: Text Should render a <Headline1 /> 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Text
      accessibilityLabel="txt-test-h1"
      accessible={true}
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "System",
            "fontWeight": "400",
            "textAlign": "left",
          },
          Array [
            Object {
              "color": "#000000",
              "fontSize": 12,
            },
            Object {
              "fontFamily": "OpenSans",
              "fontSize": 24,
              "lineHeight": 28,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            false,
            undefined,
            undefined,
            false,
            false,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            false,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
          ],
        ]
      }
      testID="txt-test-h1"
    >
      Test h1
    </Text>
  </View>
</View>
`;

exports[`Component: Text Should render a <Label /> 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Text
      accessibilityLabel="txt-test-label"
      accessible={true}
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "System",
            "fontWeight": "400",
            "textAlign": "left",
          },
          Array [
            Object {
              "color": "#000000",
              "fontSize": 12,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            Object {
              "fontFamily": "OpenSans",
              "fontSize": 8,
              "lineHeight": 9,
            },
            false,
            undefined,
            undefined,
            false,
            false,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            false,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
          ],
        ]
      }
      testID="txt-test-label"
    >
      Test label
    </Text>
  </View>
</View>
`;

exports[`Component: Text Should render a <Subtitle /> 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Text
      accessibilityLabel="txt-test-subtitle"
      accessible={true}
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "System",
            "fontWeight": "400",
            "textAlign": "left",
          },
          Array [
            Object {
              "color": "#000000",
              "fontSize": 12,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            Object {
              "fontFamily": "OpenSans",
              "fontSize": 12,
              "lineHeight": 14,
            },
            undefined,
            undefined,
            false,
            undefined,
            undefined,
            false,
            false,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            false,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
          ],
        ]
      }
      testID="txt-test-subtitle"
    >
      Test subtitle
    </Text>
  </View>
</View>
`;
